File: backend/daemon-ts/src/ledger/cache_forensics.ts
/**
 * Prompt-cache forensic JSONL log.
 *
 * Mirrors the Rust `shore_llm::cache_forensics` response-side behavior:
 * best-effort append-only writes to `{cache_dir}/cache_forensics.jsonl`.
 * Request-side breakpoint logging can be added once the TS request mirror
 * grows a stable placement hook.
 */

import fs from "node:fs";
import path from "node:path";

export interface ResponseLog {
  callId: number;
  model: string;
  character: string;
  callType: string;
  inputTokens: number;
  outputTokens: number;
  cacheReadTokens: number;
  cacheCreationTokens: number;
}

export interface RequestLog {
  callId: number;
  character?: string;
  model: string;
  msgCount: number;
  msgBreakpoints: number[];
  sysBreakpoints: number[];
  sysBlocks: number;
  prefixHash: string;
  hasExistingMarkers: boolean;
  cacheEnabled: boolean;
  rid?: string;
}

export interface ErrorLog {
  callId: number;
  model: string;
  character: string;
  callType: string;
  error: string;
}

export class CacheForensics {
  private nextId = 0;
  private readonly filePath: string;

  private constructor(cacheDir: string) {
    fs.mkdirSync(cacheDir, { recursive: true });
    this.filePath = path.join(cacheDir, "cache_forensics.jsonl");
  }

  static open(cacheDir: string): CacheForensics {
    return new CacheForensics(cacheDir);
  }

  nextCallId(): number {
    const id = this.nextId;
    this.nextId += 1;
    return id;
  }

  logRequest(entry: RequestLog): void {
    this.writeEntry({
      ts: rfc3339LocalNow(),
      type: "request",
      call_id: entry.callId,
      character: entry.character ?? null,
      model: entry.model,
      msg_count: entry.msgCount,
      msg_breakpoints: entry.msgBreakpoints,
      sys_breakpoints: entry.sysBreakpoints,
      sys_blocks: entry.sysBlocks,
      prefix_hash: entry.prefixHash,
      has_existing_markers: entry.hasExistingMarkers,
      cache_enabled: entry.cacheEnabled,
      rid: entry.rid ?? null,
    });
  }

  logResponse(entry: ResponseLog): void {
    this.writeEntry({
      ts: rfc3339LocalNow(),
      type: "response",
      call_id: entry.callId,
      model: entry.model,
      character: entry.character,
      call_type: entry.callType,
      input_tokens: entry.inputTokens,
      output_tokens: entry.outputTokens,
      cache_read_tokens: entry.cacheReadTokens,
      cache_creation_tokens: entry.cacheCreationTokens,
    });
  }

  logError(entry: ErrorLog): void {
    this.writeEntry({
      ts: rfc3339LocalNow(),
      type: "error",
      call_id: entry.callId,
      model: entry.model,
      character: entry.character,
      call_type: entry.callType,
      error: entry.error,
    });
  }

  private writeEntry(entry: Record<string, unknown>): void {
    try {
      fs.appendFileSync(this.filePath, `${JSON.stringify(entry)}\n`, "utf8");
    } catch {
      // Diagnostic-only. Never fail the main generation path on I/O.
    }
  }
}

function rfc3339LocalNow(): string {
  const now = new Date();
  const tzOffsetMinutes = -now.getTimezoneOffset();
  const sign = tzOffsetMinutes >= 0 ? "+" : "-";
  const abs = Math.abs(tzOffsetMinutes);
  const tzh = String(Math.floor(abs / 60)).padStart(2, "0");
  const tzm = String(abs % 60).padStart(2, "0");
  const pad = (n: number, w = 2): string => String(n).padStart(w, "0");
  const ms = String(now.getMilliseconds()).padStart(3, "0");
  return (
    `${now.getFullYear()}-${pad(now.getMonth() + 1)}-${pad(now.getDate())}` +
    `T${pad(now.getHours())}:${pad(now.getMinutes())}:${pad(now.getSeconds())}.${ms}${sign}${tzh}:${tzm}`
  );
}

File: backend/llm/src/cache_forensics.rs
//! Cache forensics — on-disk diagnostic log for every Anthropic cache event.
//!
//! Writes append-only JSONL to `{cache_dir}/cache_forensics.jsonl`.
//! Enabled at startup via `enable()`.  Each Anthropic API call produces
//! a "request" entry (breakpoint placement) and, if the response has
//! cache_creation_tokens > 0, a "cache_write" entry.  Both share the
//! same `call_id` for correlation.

use std::io::Write;
use std::path::PathBuf;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::OnceLock;

use serde_json::json;

static FORENSIC_DIR: OnceLock<PathBuf> = OnceLock::new();
static CALL_COUNTER: AtomicU64 = AtomicU64::new(0);

#[derive(Debug, Clone, Copy)]
pub struct RequestLog<'a> {
    pub call_id: u64,
    pub character: Option<&'a str>,
    pub model: &'a str,
    pub msg_count: usize,
    pub msg_breakpoints: &'a [usize],
    pub sys_breakpoints: &'a [usize],
    pub sys_blocks: usize,
    pub prefix_hash: u64,
    pub has_existing_markers: bool,
    pub cache_enabled: bool,
    pub rid: Option<&'a str>,
}

#[derive(Debug, Clone, Copy)]
pub struct ResponseLog<'a> {
    pub call_id: u64,
    pub model: &'a str,
    pub character: &'a str,
    pub call_type: &'a str,
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_read_tokens: u64,
    pub cache_creation_tokens: u64,
}

/// Enable cache forensics.  Call once at startup with the cache directory.
pub fn enable(cache_dir: PathBuf) {
    let _ignored = FORENSIC_DIR.set(cache_dir);
}

/// Whether forensics is enabled.
pub fn is_enabled() -> bool {
    FORENSIC_DIR.get().is_some()
}

/// Allocate a monotonic call ID for correlating request/response entries.
pub fn next_call_id() -> u64 {
    CALL_COUNTER.fetch_add(1, Ordering::Relaxed)
}

/// Write an arbitrary JSON entry to the forensic log.
pub fn write_entry(entry: &serde_json::Value) {
    let Some(dir) = FORENSIC_DIR.get() else {
        return;
    };
    let path = dir.join("cache_forensics.jsonl");
    let line = entry.to_string();

    // Best-effort append — don't crash or block the main path on I/O errors.
    if let Ok(mut f) = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&path)
    {
        let _ignored = writeln!(f, "{line}");
    }
}

/// Log the request-side cache placement for an Anthropic call.
pub fn log_request(entry: RequestLog<'_>) {
    let ts = chrono::Local::now().to_rfc3339();
    write_entry(&json!({
        "ts": ts,
        "type": "request",
        "call_id": entry.call_id,
        "character": entry.character,
        "model": entry.model,
        "msg_count": entry.msg_count,
        "msg_breakpoints": entry.msg_breakpoints,
        "sys_breakpoints": entry.sys_breakpoints,
        "sys_blocks": entry.sys_blocks,
        "prefix_hash": format!("{:016x}", entry.prefix_hash),
        "has_existing_markers": entry.has_existing_markers,
        "cache_enabled": entry.cache_enabled,
        "rid": entry.rid,
    }));
}

/// Log the response-side cache event.
pub fn log_response(entry: ResponseLog<'_>) {
    let ts = chrono::Local::now().to_rfc3339();
    write_entry(&json!({
        "ts": ts,
        "type": "response",
        "call_id": entry.call_id,
        "model": entry.model,
        "character": entry.character,
        "call_type": entry.call_type,
        "input_tokens": entry.input_tokens,
        "output_tokens": entry.output_tokens,
        "cache_read_tokens": entry.cache_read_tokens,
        "cache_creation_tokens": entry.cache_creation_tokens,
    }));
}

/// Log a failed API call (error from generate/stream) so keepalive and
/// other failures are visible in the forensic log, not just journald.
pub fn log_error(call_id: u64, model: &str, character: &str, call_type: &str, error: &str) {
    let ts = chrono::Local::now().to_rfc3339();
    write_entry(&json!({
        "ts": ts,
        "type": "error",
        "call_id": call_id,
        "model": model,
        "character": character,
        "call_type": call_type,
        "error": error,
    }));
}

/// Fire a desktop notification for a cache anomaly.
///
/// Spawns `notify-send` in the background — best-effort, never blocks.
pub fn notify_anomaly(
    character: &str,
    anomaly: &str,
    call_type: &str,
    cache_read: u64,
    cache_write: u64,
) {
    if !is_enabled() {
        return;
    }
    let summary = format!("shore: cache {anomaly}");
    let body = format!("{character} ({call_type})\nread={cache_read} write={cache_write}");
    let _ignored = std::process::Command::new("notify-send")
        .args(["--urgency=normal", "--app-name=shore", &summary, &body])
        .stdin(std::process::Stdio::null())
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .spawn();
}
